docs/dev/alternative_release_image_sources.md

```diff
@@ -0,0 +1,151 @@
+# Alternative Release-Image Sources
+
+## release-image content vs release-image source
+
+`release-image` content is what operators get installed and what version of each operator (i.e. the container image) gets installed to the cluster. While, `release-image` source is **where** the `release-image` content gets pulled.
+
+Currently the installer controls **both** the `release-image` content and the `release-image` source using the embedded release-image location or the `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE` env, but to allow users to use a release-image from their private registry (change the `release-image` source), but keep the `release-image` content identical to one vetted by OpenShift, the installer needs to allow specifying source separate from the content.
+
+## Controlling the content
+
+The content of the `release-image`, i.e. the digest, continues to be controlled by the embedded release-image location or the `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE` env.
+
+## Controlling the source
+
+The installer allows the users to specify the sources for the release-image repository and the other repositories referenced in the release-image through the InstallConfig.
+
+## Details
+
+The design is based on the assumption that all flows of using multiple sources/repositories for the release-image originate from user's using the `oc adm release mirror` command to create those sources/repositories.
+
+## InstallConfig
+
+```go
+type InstallConfig struct {
+    ///
+
+    // ReleaseImageSources is the list of sources/repositories for pulling the release-image
+    // content.
+    // Any of the source/repository can be used to pull the release-image content.
+    // The order is the list has no meaning in terms of preference.
+    ReleaseImageSources []ReleaseImageSource `json:"releaseImageSources"`
```

Comment:
So a key question here is whether we want people to provide initial content mirrors here.  I think a reasonable person could say this as cluster level setting (this set of content from A, this set of content for B).  I'm concerned that this is too specific (looks different from ICSP in a significant way).  

As an alternative, if we called this InstallContentSources and the structure matched RepositoryDigestMirrors, what is the impact?  This would be the initial set of ICSP objects - it would expose a subset of the fields there (anyone who wants more can always add their own ICSP, but that's a 10% case at best), and you should use this field to seed your initial mirrors for core cluster content.

I'm concerned that we may have a case where core content needs two sets of mirrors at some point (release digests A-J come from repository 1, digests K-Z come from repository 2), which would necessitate the array of arrays struct like we have for ICSP.  Also, today we have the samples operator (which in theory should be in OLM, but isn't, and won't be for a while) which won't come up properly unless you also mirror it.

Are we 100% positive we only need one mirror list, vs potentially needing 2 at any point.  That's the argument for array or array of arrays.

Reply:
In general, looking more like ICSP (but with a better name and all extraneous stuff scraped out) lets us explain it better.

I.e.

```
InstallContentSources []InstallContentSource

type InstallContentSource struct {
  // only fields from ICSP.RepositoryDigestMirrors that make sense
  Sources []string
}
```

Reply:
> As an alternative, if we called this InstallContentSources and the structure matched RepositoryDigestMirrors, what is the impact?

The code would have to handle input where the two `InstallContentSources` arrays had common elements. (This is not a blocker - the code can either explicitly reject such input and tell the user to combine the lists, or use the code that already exists in machine-config-operator to combine them; just something to pay attention to.)

Reply:
>  (release digests A-J come from repository 1, digests K-Z come from repository 2), which would necessitate the array of arrays struct like we have for ICSP.

with current ICSP api definition `[a,b,c] [c,d,e] => [a,b,c,d]` so how does multiple lists help with
`[release, release-image-refA-J, mirror] [release, release-image-refK-Z, mirror]` in the installer as that already means `[release, release-image-refA-J, release-image-refK-Z, mirror]`

@smarterclayton I don't completely understand the use-case for core-content.

and for extra content sources, i'm more incline to push that to day-2.. no reason that needs to be day-1